### PR TITLE
feat(http.Agent): allow passing agent option as an object

### LINKED ISSUE #15: Allow to pass all available https options
Currently only a subset of all the available options are supported https://github.com/bitinn/node-fetch/blob/master/index.js#L66-L80

For `https` important options are missing (e.g.: `rejectUnauthorized`, `secureProtocol`, `key`, `cert`, ...)
https://nodejs.org/api/https.html#https_https_request_options_callback

I should be able to pass all of those options.

### PR DESCRIPTION
**Pull request by ahmadnassri**
allows for controlling `http.Agent` without manually creating a new external object, this is useful

to prevent "requiring" `http`/`https` in SPA and isomorphic apps

Fixes #15

### CODE REVIEW

File: test/test.js [hunk 1]
```diff
@@ -1319,6 +1319,20 @@ describe('node-fetch', () => {
 		});
 	});
 
+	it('should create http.Agent class if options.agent is provided as an object', function() {
```
Comment (TimothyGu) [inline test/test.js]: Is there a way to test `https.Agent` also?

Comment (ahmadnassri) [inline test/test.js]: yes, we can add an https specific test. 

I didn't think it was necessary since `https.Agent` inherits `http.Agent` and the logic is mainly concerned with initializing the object internally.

can add a test later today.

Comment (ahmadnassri) [inline test/test.js]: upon further study, I'm sticking with my initial comment, since `https.agent` inherits `http.agent` and the added logic is only concerned with initializing the objects.

I do think adding an `https.agent` test in general is great, but that is outside the scope of this PR and can be made into a separate branch & PR

Comment (TimothyGu) [inline test/test.js]: @ahmadnassri, agreed. I'll try to merge this PR later today. Thanks for your work!

Comment (ahmadnassri) [inline test/test.js]: thanks!

**Comment by TimothyGu** [comment]
@ahmadnassri, actually, on a second thought I'm not sure if this is the best idea. After this patch, if an options object is passed, a new agent will be created on every fetch. I don't think we should encourage that.

**Comment by ahmadnassri** [comment]
@TimothyGu I could see that concern. However it is also an optimization that can be left in user-land.

this change does not prevent performancr aware users from creating and using a global agent.

also, by creating different agents for each request it facilitates more user-level, fine-tune grained control, that would otherwise not be possible with a global agent.

**Comment by TimothyGu** [comment]
Rebased and applied in ec29e3d2640586bc07cddf894e2aa10461b20ec0. Thanks!

**Comment by Jakobud** [comment]
So, this is an old PR, but I see this was reverted. So does this no longer work? I can see that you can still pass an `agent` option into `fetch`. What does it do? Does this still allow passing an `https.Agent()` into it?

Also regarding the above PR, instead of creating an Agent for each request, you should have just had the Agent created when instantiating the client, and then just used that Agent on each request, so you aren't make a new one each time.

**Comment by TimothyGu** [comment]
@Jakobud

> Does this still allow passing an `https.Agent()` into it?

Yes! We’ve always allowed this.

**Comment by Jakobud** [comment]
Oh okay well then what code got reverted exactly? Automatic creation of an agent if an object is passed in?

**Comment by TimothyGu** [comment]
@Jakobud 

> Automatic creation of an agent if an object is passed in?

Yes, that was reverted.

> Also, the agent is cached so that it doesn't get created over again for every request, correct?

We don't control the Agent object you pass in, so it is your responsibility to use the same Agent object yourself if you don't want to create it over again.

**Comment by faceleg** [comment]
I've just run into a problem that this PR would solve - is there anything I could do to help get this merged?

**Comment by jimmywarting** [comment]
@faceleg with v3 you can supply a function that can return an agent based on the protocol being used.
```js
const httpAgent = new http.Agent({ keepAlive: true });
const httpsAgent = new https.Agent({ keepAlive: true });

fetch(url, {
  agent (url) { 
    return url.protocol === 'https:' ? httpsAgent : httpAgent
  }
})
```
